### PR TITLE
Let revset engine create change id index

### PR DESCRIPTION
**Pull request by martinvonz**
# Checklist

If applicable:
- [ ] I have updated `CHANGELOG.md`
- [ ] I have updated the documentation (README.md, docs/, demos/)
- [ ] I have updated the config schema (src/config-schema.json)
- [ ] I have added tests to cover my changes

### CODE REVIEW

**Review by yuja — COMMENTED** [review]
I'll look into unsafe bits tomorrow.

File: lib/src/default_revset_engine.rs [hunk 1]
```diff
@@ -384,16 +384,16 @@ pub fn evaluate<'index>(
     repo: &'index dyn Repo,
     expression: &RevsetExpression,
 ) -> Result<Box<dyn Revset<'index> + 'index>, RevsetError> {
-    let revset_impl = evaluate_impl(repo, expression)?;
-    Ok(Box::new(revset_impl))
+    let internal_revset = internal_evaluate(repo, expression)?;
+    Ok(Box::new(RevsetImpl::new(internal_revset)))
```
Comment (yuja) [inline lib/src/default_revset_engine.rs]: Now we can remove `impl<'index> ToPredicateFn<'index> for RevsetImpl<'index>` (line ~70)

File: lib/src/default_index_store.rs [hunk 2]
```diff
@@ -494,6 +494,11 @@ impl MutableIndexImpl {
         }
     }
 
+    #[cfg(test)]
+    pub fn as_composite(&self) -> CompositeIndex {
+        CompositeIndex(self)
```
Comment (yuja) [inline lib/src/default_index_store.rs:499]: nit: no longer used in tests

File: lib/src/repo.rs [hunk 3]
```diff
@@ -209,21 +213,28 @@ impl ReadonlyRepo {
...
+            .get_or_init(|| {
+                let revset: Box<dyn Revset<'a>> = RevsetExpression::all().evaluate(self).unwrap();
+                let change_id_index: Box<dyn ChangeIdIndex + 'a> = revset.change_id_index();
+                let change_id_index: Box<dyn ChangeIdIndex> =
+                    unsafe { std::mem::transmute(change_id_index) };
+                change_id_index
+            })
+            .as_ref();
+        let change_id_index: &'a (dyn ChangeIdIndex + 'a) =
+            unsafe { std::mem::transmute(change_id_index) };
```
Comment (yuja) [inline lib/src/repo.rs:236]: Unneeded unsafe cast

**Review by yuja — APPROVED** [review]

File: lib/src/default_revset_engine.rs (hunk 1 above)
Comment (martinvonz) [reply lib/src/default_revset_engine.rs]: Good point. Done.

File: lib/src/default_index_store.rs (hunk 2 above)
Comment (martinvonz) [reply lib/src/default_index_store.rs]: Ah, nice. Done.

File: lib/src/repo.rs (hunk 3 above)
Comment (martinvonz) [reply lib/src/repo.rs]: Good catch. (I think this was needed before I removed the lifetime parameter to `ChangeIdIndex`.)

File: lib/src/repo.rs [hunk 4]
```diff
@@ -209,21 +213,28 @@ impl ReadonlyRepo {
...
-            )
-        })
+    fn change_id_index<'a>(&'a self) -> &'a (dyn ChangeIdIndex + 'a) {
+        let change_id_index: &'a (dyn ChangeIdIndex + 'a) = self
+            .change_id_index
+            .get_or_init(|| {
+                let revset: Box<dyn Revset<'a>> = RevsetExpression::all().evaluate(self).unwrap();
+                let change_id_index: Box<dyn ChangeIdIndex + 'a> = revset.change_id_index();
+                let change_id_index: Box<dyn ChangeIdIndex> =
+                    unsafe { std::mem::transmute(change_id_index) };
```
[the remark this answers is not in the document]
Comment (martinvonz) [reply lib/src/repo.rs:232]: Good point. I've added a comment. I hope to replace that Repo argument by an Index argument soon.

**Pull request merged**